src/java.base/share/classes/com/sun/crypto/provider/KeyWrapCipher.java

```diff
@@ -0,0 +1,753 @@
+                    outLen = padIdx;
+                }
+            } else {
+                outLen = cipher.encryptFinal(dataBuf, 0, dataIdx, null, -1);
+            }
+            // only write out the result after verification succeeds
+            System.arraycopy(dataBuf, 0, out, outOfs, outLen);
+            return outLen;
+        } catch (ShortBufferException sbe) {
+            // should never happen
+            throw new AssertionError();
+        } finally {
+            // reset dataBuf for subsequent operations
+            dataIdx = 0;
+        }
+    }
+
+    /**
+     * Returns the parameters used with this cipher.
+     *
+     * @return AlgorithmParameters object containing IV.
+     */
+    @Override
+    protected AlgorithmParameters engineGetParameters() {
+        AlgorithmParameters params = null;
+
+        byte[] iv = cipher.getIV();
+        String algo = "AES/" + cipher.getFeedback() + "/NoPadding";
+        try {
+            params = AlgorithmParameters.getInstance(algo,
+                    SunJCE.getInstance());
+            params.init(new IvParameterSpec(iv));
+        } catch (NoSuchAlgorithmException | InvalidParameterSpecException e) {
+            // should never happen
+            throw new AssertionError();
+        }
+        return params;
+    }
+
+    /**
+     * Returns the key size of the given key object in number of bits.
+     *
+     * @param key the key object.
+     *
+     * @return the "effective" key size of the given key object.
+     *
+     * @exception InvalidKeyException if <code>key</code> is invalid.
+     */
+    protected int engineGetKeySize(Key key) throws InvalidKeyException {
+        byte[] encoded = key.getEncoded();
+        if (encoded == null)  {
+            throw new InvalidKeyException("Cannot decide key length");
+        }
+
+        // only need length
+        Arrays.fill(encoded, (byte) 0);
+        int keyLen = encoded.length;
+        if (!key.getAlgorithm().equalsIgnoreCase("AES") ||
+            !AESCrypt.isKeySizeValid(keyLen) ||
+            (fixedKeySize != -1 && fixedKeySize != keyLen)) {
+            throw new InvalidKeyException("Invalid key length: " +
+                    keyLen + " bytes");
+        }
+        return Math.multiplyExact(keyLen, 8);
+    }
+
+    /**
+     * Wrap a key.
+     *
+     * @param key the key to be wrapped.
+     *
+     * @return the wrapped key.
+     *
+     * @exception IllegalBlockSizeException if this cipher is a block
+     * cipher, no padding has been requested, and the length of the
+     * encoding of the key to be wrapped is not a
+     * multiple of the block size.
+     *
+     * @exception InvalidKeyException if it is impossible or unsafe to
+     * wrap the key with this cipher (e.g., a hardware protected key is
+     * being passed to a software only cipher).
+     */
+    @Override
+    protected byte[] engineWrap(Key key)
+            throws IllegalBlockSizeException, InvalidKeyException {
```

Comment:
Is it okay that we can call `cipher.init(Cipher.ENCRYPT_MODE, ...)` and then use it with `cipher.wrap()`?

Also, can we simply delegate the heavy lifting to `engineDoFinal()` (or the new suggested helper method) rather than duplicating this logic here?

Reply:
Right, I will add additional checking on the Cipher.XXX_MODE. Agree with you regarding reducing the code duplication.